File: server/domain/trade_filter.go
package domain

import "time"

type TradeFilter struct {
	// ID/チケット番号フィルター
	TicketIDs []int `json:"ticketIds,omitempty"`

	// 期間フィルター（取引開始・終了）
	StartDate *time.Time `json:"startDate,omitempty"`
	EndDate   *time.Time `json:"endDate,omitempty"`

	// 取引種別フィルター
	Types []string `json:"types,omitempty"` // 例: ["buy", "sell"]

	// 通貨ペア/商品フィルター
	Items []string `json:"items,omitempty"` // 例: ["usdjpy", "eurusd"]

	// 数値範囲フィルター
	SizeMin      *float64 `json:"sizeMin,omitempty"`
	SizeMax      *float64 `json:"sizeMax,omitempty"`
	ProfitMin    *float64 `json:"profitMin,omitempty"`
	ProfitMax    *float64 `json:"profitMax,omitempty"`
	OpenPriceMin *float64 `json:"openPriceMin,omitempty"`
	OpenPriceMax *float64 `json:"openPriceMax,omitempty"`

	// ページング
	Page     int `json:"page,omitempty"`
	PageSize int `json:"pageSize,omitempty"`

	// ソート
	SortBy    string `json:"sortBy,omitempty"`
	SortOrder string `json:"sortOrder,omitempty"` // "asc" or "desc"
}

File: utils/tradeFilter.ts
import { PAGINATION } from '@/constants/pagination';
import { TradeFilter } from '@/types/trade';

/**
 * TradeFilterをAPI用に正規化する共通関数
 * - type=allならtypesはundefined
 * - typeがbuy/sellなら[それ]
 * - itemsはそのまま配列で扱う
 * - 日付はJST 0:00:00/23:59:59.999でISO8601文字列
 */
export function buildTradeFilterParams(userFilter: TradeFilter): TradeFilter {
  // 日付をJST 0:00:00/23:59:59.999でISO8601
  let startDate: Date | undefined = undefined;
  let endDate: Date | undefined = undefined;

  if (userFilter.startDate) {
    startDate = new Date(userFilter.startDate);
    startDate.setHours(0, 0, 0, 0);
  }

  if (userFilter.endDate) {
    endDate = new Date(userFilter.endDate);
    endDate.setHours(23, 59, 59, 999);
  }

  return {
    ...userFilter,
    startDate,
    endDate,
    sortBy: userFilter.sortBy || 'openTime',
    sortOrder: userFilter.sortOrder || 'desc',
    page: userFilter.page || PAGINATION.DEFAULT_PAGE,
    pageSize: userFilter.pageSize || PAGINATION.DEFAULT_PAGE_SIZE
  };
}

/**
 * AIアシスタントのツール呼び出し用フィルターパラメータを生成する関数
 * @param aiParams 期間パラメータ（startDate, endDate）
 * @returns AI function calling用のフィルターパラメータ
 */
export function buildAIParamsFilter(
  aiParams: { startDate?: string; endDate?: string }
) {
  // 日付をJST 0:00:00/23:59:59.999でISO8601
  const startDate: Date | undefined = aiParams.startDate ? new Date(aiParams.startDate) : undefined;
  const endDate: Date | undefined = aiParams.endDate ? new Date(aiParams.endDate) : undefined;

  if (startDate) {
    startDate.setHours(0, 0, 0, 0);
  }

  if (endDate) {
    endDate.setHours(23, 59, 59, 999);
  }

  return {
    startDate,
    endDate,
  };
}
